core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -543,62 +534,95 @@ public IndexResult index(Index index) throws IOException {
+                        versionForIndexing = 1L;
+                        currentNotFoundOrDeleted = true;
+                        seqNoForIndexing = seqNoService().generateSeqNo();
+                        performOperation = true;
                     } else {
-                        currentVersion = checkDeletedAndGCed(versionValue);
-                        deleted = versionValue.delete();
+                        // resolves incoming version
+                        final VersionValue versionValue = resolveDocVersion(index);
+                        final long currentVersion;
+                        if (versionValue == null) {
+                            currentVersion = Versions.NOT_FOUND;
+                            currentNotFoundOrDeleted = true;
+                            useLuceneUpdateDocument = false;
+                        } else {
+                            currentVersion = versionValue.getVersion();
+                            currentNotFoundOrDeleted = versionValue.isDelete();
+                            useLuceneUpdateDocument = versionValue.isDelete() == false; // last operation is not delete
+                        }
+                        if (index.versionType().isVersionConflictForWrites(currentVersion, index.version(), currentNotFoundOrDeleted)) {
+                            earlyResultOnPreFlightError = Optional.of(new IndexResult(
+                                new VersionConflictEngineException(shardId, index, currentVersion, currentNotFoundOrDeleted),
+                                currentVersion
+                            ));
+                            versionForIndexing = Versions.NOT_FOUND;
+                            seqNoForIndexing = SequenceNumbersService.UNASSIGNED_SEQ_NO;
+                            performOperation = false;
+                        } else {
+                            versionForIndexing = index.versionType().updateVersion(currentVersion, index.version());
+                            seqNoForIndexing = seqNoService().generateSeqNo();
+                            performOperation = true;
+                        }
                     }
-                }
-                final long expectedVersion = index.version();
-                Optional<IndexResult> resultOnVersionConflict;
-                try {
-                    final boolean isVersionConflict = checkVersionConflict(index, currentVersion, expectedVersion, deleted);
-                    resultOnVersionConflict = isVersionConflict ? Optional.of(new IndexResult(currentVersion, index.seqNo(), false))
-                        : Optional.empty();
-                } catch (IllegalArgumentException | VersionConflictEngineException ex) {
-                    resultOnVersionConflict = Optional.of(new IndexResult(ex, currentVersion, index.seqNo()));
+                } else if (canOptimizeAddDocument && mayHaveBeenIndexedBefore(index) == false) {
```

Comment:
can we fold this into an `else {` and start again with `if` in there with a comment that we are now on a replica? it would read cleaner